pkgs/development/tools/metals/default.nix

```diff
@@ -0,0 +1,52 @@
+{ stdenv, lib, fetchurl, coursier, jdk, jre, makeWrapper }:
+
+let
+  baseName = "metals";
+  version = "0.3.2";
+  deps = stdenv.mkDerivation {
+    name = "${baseName}-deps-${version}";
+    buildCommand = ''
+      export COURSIER_CACHE=$(pwd)
+      ${coursier}/bin/coursier fetch org.scalameta:metals_2.12:${version} \
+        -r "sonatype:releases" > deps
+      mkdir -p $out/share/java
+      cp -n $(< deps) $out/share/java/
+    '';
+    outputHashMode = "recursive";
+    outputHashAlgo = "sha256";
+    outputHash     = "12hvnvrm63z2ibwzkr3c9fbx32dfkcfvp1pmmc3s6jqfjl5wnlgm";
+  };
+in
+stdenv.mkDerivation {
+  name = "${baseName}-${version}";
+  version = version;
+
+  buildInputs = [ jdk makeWrapper deps ];
+
+  phases = [ "installPhase" ];
+
+  installPhase = ''
+    mkdir -p $out/bin
+
+    makeWrapper ${jre}/bin/java $out/bin/metals \
+      --prefix PATH : ${lib.makeBinPath [ jdk ]} \
+      --add-flags "-cp $CLASSPATH scala.meta.metals.Main" \
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = https://scalameta.org/metals/;
+    license = licenses.asl20;
+    description = "Language server for Scala";
+    longDescription = ''
+      This executable includes a minimal number of flags by default, and when
+      using it, you will likely want to pass in additional flags such as:
```

Comment:
I just realized that if you add additional arguments to the command as its currently set up, they will be interpreted as arguments to the `main` method as opposed to as java flags (which is what the call probably wants). I'm going to close this out until we resolve this.

Reply:
It's not a big change in Metals to support regular cli argument `-Dmetals.client=emacs`. However, users should still be able to customize `-Xmx` and `-Xms` in my opinion.

Reply:
@olafurpg I agree. I don't see much benefit in changing metals to support the CLI argument. That might just serve to make it more confusing.

Reply:
One mechanism is similar to @ceedubs referenced: Add a config param to the function that produces the derivation. When called using `callPackage` (as is the standard in nixpkgs) This can be overloaded in nixpkgs using the `override` function:

* https://nixos.org/nixpkgs/manual/index.html#sec-pkg-override